File: mage/src/main/java/mil/nga/giat/mage/login/idp/IdpLoginActivity.java
package mil.nga.giat.mage.login.idp;

import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.view.View;
import android.webkit.JavascriptInterface;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;

import androidx.fragment.app.FragmentActivity;

import com.google.gson.JsonObject;

import mil.nga.giat.mage.R;
import mil.nga.giat.mage.disclaimer.DisclaimerActivity;
import mil.nga.giat.mage.event.EventsActivity;
import mil.nga.giat.mage.login.LoginActivity;
import mil.nga.giat.mage.login.SignupActivity;
import mil.nga.giat.mage.sdk.login.AbstractAccountTask;
import mil.nga.giat.mage.sdk.login.AccountDelegate;
import mil.nga.giat.mage.sdk.login.AccountStatus;
import mil.nga.giat.mage.sdk.login.IdpLoginTask;
import mil.nga.giat.mage.sdk.login.IdpSignupTask;
import mil.nga.giat.mage.sdk.utils.DeviceUuidFactory;

/**
 * Created by wnewman on 10/14/15.
 */
public class IdpLoginActivity extends FragmentActivity implements AccountDelegate {

    public enum IdpType {
        SIGNIN,
        SIGINUP
    }

    private static final String LOG_NAME = IdpLoginActivity.class.getName();

    public static final String EXTRA_SERVER_URL = "EXTRA_SERVER_URL";
    public static final String EXTRA_IDP_TYPE = "EXTRA_IDP_TYPE";
    public static final String EXTRA_IDP_STRATEGY = "EXTRA_IDP_STRATEGY";

    private String serverURL;
    private String idpURL;
    private IdpType idpType;
    private String idpStrategy;

    private WebView webView;
    private View progress;
    private String uuid;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_idp);

        serverURL = getIntent().getStringExtra(EXTRA_SERVER_URL);
        idpType = (IdpType) getIntent().getSerializableExtra(EXTRA_IDP_TYPE);
        idpStrategy = getIntent().getStringExtra(EXTRA_IDP_STRATEGY);
        idpURL = String.format("%s/auth/%s/signin", serverURL, idpStrategy);

        uuid = new DeviceUuidFactory(this).getDeviceUuid().toString();

        progress = findViewById(R.id.progress);

        webView = findViewById(R.id.webView);
        WebSettings settings = webView.getSettings();
        settings.setJavaScriptEnabled(true);
        settings.setJavaScriptCanOpenWindowsAutomatically(true);
        webView.addJavascriptInterface(this, "Android");
        webView.setWebViewClient(new WebViewClient() {
            public void onPageFinished(WebView view, String url) {
                progress.setVisibility(View.INVISIBLE);

                 if (url.contains("/callback")) {
                    webView.setVisibility(View.INVISIBLE);
                    webView.loadUrl("javascript:Android.getLogin(JSON.stringify(login));");
                }
            }
        });

        webView.loadUrl(idpURL);
    }

    @JavascriptInterface
    public void getLogin(String login) {
        AbstractAccountTask loginTask = idpType == IdpType.SIGNIN ?
                new IdpLoginTask(this, getApplicationContext()) :
                new IdpSignupTask(this, getApplicationContext());

        loginTask.execute(idpStrategy, uuid, login );
    }

    @Override
    public void finishAccount(AccountStatus accountStatus) {
        if (accountStatus.getStatus().equals(AccountStatus.Status.SUCCESSFUL_LOGIN)) {
            SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(this);
            boolean showDisclaimer = sharedPreferences.getBoolean(getString(R.string.serverDisclaimerShow), false);

            Intent intent = showDisclaimer ?
                    new Intent(getApplicationContext(), DisclaimerActivity.class) :
                    new Intent(getApplicationContext(), EventsActivity.class);

            startActivity(intent);
            finish();
        } else if (accountStatus.getStatus().equals(AccountStatus.Status.SUCCESSFUL_REGISTRATION)) {
            Intent intent = new Intent();
            intent.putExtra(LoginActivity.EXTRA_IDP_UNREGISTERED_DEVICE, true);
            setResult(RESULT_OK, intent);
            finish();
        } else if (accountStatus.getStatus().equals(AccountStatus.Status.SUCCESSFUL_SIGNUP)) {
            Intent intent = new Intent();
            setResult(RESULT_OK, intent);
            finish();
        } else if (accountStatus.getStatus().equals(AccountStatus.Status.FAILED_SIGNUP)) {
            JsonObject json = accountStatus.getAccountInformation();
            String errorMessage = "Your account could not be created, please contact your MAGE administrator";

            if (json.has("errorMessage")) {
                errorMessage = json.get("errorMessage").getAsString();
            }

            Intent intent = new Intent();
            intent.putExtra(SignupActivity.EXTRA_IDP_ERROR, true);
            intent.putExtra(SignupActivity.EXTRA_IDP_ERROR_MESSAGE, errorMessage);
            setResult(RESULT_OK, intent);
            finish();
        } else {
            Intent intent = new Intent();
            intent.putExtra(LoginActivity.EXTRA_IDP_ERROR, true);
            setResult(RESULT_OK, intent);
            finish();
        }
    }
}

File: mage/src/main/java/mil/nga/giat/mage/login/idp/IdpLoginActivity.kt
package mil.nga.giat.mage.login.idp

import android.content.Context
import android.content.Intent
import android.graphics.Bitmap
import android.net.Uri
import android.os.Bundle
import android.util.TypedValue
import androidx.appcompat.content.res.AppCompatResources
import androidx.browser.customtabs.CustomTabsIntent
import androidx.core.graphics.drawable.toBitmap
import androidx.fragment.app.FragmentActivity
import mil.nga.giat.mage.R
import mil.nga.giat.mage.login.AccountStateActivity

/**
 * Created by wnewman on 10/14/15.
 */
class IdpLoginActivity : FragmentActivity() {
    companion object {
        private const val KEY_AUTHORIZATION_STARTED = "AUTHORIZATION_STARTED"
        private const val EXTRA_SERVER_URL = "EXTRA_SERVER_URL"
        private const val EXTRA_IDP_STRATEGY = "EXTRA_IDP_STRATEGY"

        const val EXTRA_IDP_TOKEN = "EXTRA_IDP_TOKEN"

        fun intent(context: Context?, url: String, strategy: String?): Intent {
            val intent = Intent(context, IdpLoginActivity::class.java)
            intent.putExtra(EXTRA_SERVER_URL, url)
            intent.putExtra(EXTRA_IDP_STRATEGY, strategy)
            return intent
        }
    }

    private var authorizationStarted = false
    private var serverURL: String? = null
    private var idpURL: String? = null
    private var  idpStrategy: String? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_idp)

        serverURL = intent.getStringExtra(EXTRA_SERVER_URL)
        idpStrategy = intent.getStringExtra(EXTRA_IDP_STRATEGY)
        idpURL = String.format("%s/auth/%s/signin?state=mobile", serverURL, idpStrategy)

        if (savedInstanceState == null) {
            extractState(intent.extras)
        } else {
            extractState(savedInstanceState)
        }
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        setIntent(intent)
    }

    override fun onResume() {
        super.onResume()

        if (!authorizationStarted) {
            val customTabsIntent = CustomTabsIntent.Builder()
                    .setCloseButtonIcon(backButtonIcon())
                    .setToolbarColor(primaryColor())
                    .build()

            customTabsIntent.launchUrl(this, Uri.parse(idpURL))

            authorizationStarted = true

            return
        }

        if (intent.data != null) {
            handleAuthenticationComplete()
        } else {
            handlAuthenticationCanceled()
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putBoolean(KEY_AUTHORIZATION_STARTED, authorizationStarted)
    }

    private fun extractState(state: Bundle?) {
        if (state == null) {
            finish()
            return
        }

        authorizationStarted = state.getBoolean(KEY_AUTHORIZATION_STARTED, false)
    }

    private fun handleAuthenticationComplete() {
        when (intent.data?.path) {
            "/invalid_account" -> {
                val intent = AccountStateActivity.intent(this, intent.data)
                startActivity(intent)
                finish()
            }
            "/authentication" -> {
                val jwt = intent.data?.getQueryParameter("token") ?: ""
                val intent = Intent()
                intent.putExtra(EXTRA_IDP_TOKEN, jwt)
                setResult(RESULT_OK, intent)
                finish()
            }
            else -> {
                handlAuthenticationCanceled()
            }
        }
    }

    private fun handlAuthenticationCanceled() {
        finish()
    }

    private fun primaryColor(): Int {
        val typedValue = TypedValue()
        theme.resolveAttribute (R.attr.colorPrimary, typedValue, true)
        return typedValue.data
    }

    private fun backButtonIcon(): Bitmap {
        return AppCompatResources.getDrawable(applicationContext, R.drawable.ic_arrow_back_white_24dp)!!.toBitmap()
    }
}